### PR TITLE
Schema Cache Improvement 2

### PR DESCRIPTION
**Pull request by dplewis**
Reduce the number of create/update/destroy object calls to cache.

After the class and fields have been validated and cached, no need to reload the schema from cache if already validated.

Create/Update/Destroy just does lookup from the cache.

I’m more than happy to answer any questions or would like feedback.

### CODE REVIEW

File: spec/RedisCacheAdapter.spec.js [hunk 1]
```diff
@@ -175,20 +175,20 @@ describe_only(() => {
 
   beforeEach(async () => {
     await cacheAdapter.clear();
-    getSpy = spyOn(cacheAdapter, 'get').and.callThrough();
-    putSpy = spyOn(cacheAdapter, 'put').and.callThrough();
     await reconfigureServer({
       cacheAdapter,
       enableSingleSchemaCache: true,
     });
+    getSpy = spyOn(cacheAdapter, 'get').and.callThrough();
```
Comment (acinader) [inline spec/RedisCacheAdapter.spec.js]: if you switch from fat arrow functions `() => {}` to `function ()` for all of all `desribe`, `it`, `beforeEach`, then you can `this.getSpy = spyOn(....`

And then use them in your tests like `expect(this.getSpy....`

not asking you to do it, just the patterns i use.

File: spec/RedisCacheAdapter.spec.js [hunk 2]
```diff
@@ -214,7 +214,57 @@ describe_only(() => {
...
+    for (let i = 0; i < 10; i++) {
+      const object = new TestObject();
+      object.set('number', i);
+      objects.push(object);
+    }
+    await Parse.Object.saveAll(objects, { batchSize: 5 });
+    expect(getSpy.calls.count()).toBe(12);
+    expect(putSpy.calls.count()).toBe(5);
+
+    getSpy.calls.reset();
```
Comment (acinader) [inline spec/RedisCacheAdapter.spec.js]: should just be able to call `reset()` in the `beforeEach`, right?

File: src/rest.js [hunk 3]
```diff
@@ -151,8 +152,10 @@ function del(config, auth, className, objectId) {
         return;
       }
     })
-    .then(() => {
-      var options = {};
+    .then(() => config.database.loadSchema())
+    .then(s => {
+      schemaController = s;
```
Comment (acinader) [inline src/rest.js]: why?  why not just `((schemaController) => {`

File: src/rest.js [hunk 4]
```diff
@@ -166,20 +169,19 @@ function del(config, auth, className, objectId) {
...
-      config.database.loadSchema().then(schemaController => {
-        const perms = schemaController.getClassLevelPermissions(className);
-        config.liveQueryController.onAfterDelete(
-          className,
-          inflatedObject,
-          null,
-          perms
-        );
-      });
+      const perms = schemaController.getClassLevelPermissions(className);
```
Comment (acinader) [inline src/rest.js]: you sure ;)

Comment (dplewis) [reply src/rest.js]: At this point the schemaData is loaded in the schemaController. The destroy call before it just deletes objects and the className is still in the schemaData

File: spec/RedisCacheAdapter.spec.js (hunk 2 above)
Comment (dplewis) [reply spec/RedisCacheAdapter.spec.js]: I added these here to isolate get and put calls before each operation. Without it is difficult to know what calls are coming from which operation.

File: src/rest.js (hunk 3 above)
Comment (dplewis) [reply src/rest.js]: Need to save a reference to be used below.

**Pull request merged**